File: src/main/java/org/nyxlang/interpreter/visitor/VarAssignVisitor.java
package org.nyxlang.interpreter.visitor;

import org.nyxlang.interpreter.IInterpreter;
import org.nyxlang.interpreter.exception.VisitorException;
import org.nyxlang.parser.ast.INode;
import org.nyxlang.parser.ast.VarAssignNode;


/**
 * Interprets the var assign node.
 */
public class VarAssignVisitor implements IVisitor {

    @Override
    public boolean matches(INode node) {
        return VarAssignNode.class == node.getClass();
    }

    @Override
    public String visit(IInterpreter interpreter, INode node) throws VisitorException {
        var varAssignNode = (VarAssignNode) node;
        interpreter.getGlobalMemory().set(
                varAssignNode.getIdentifier(),
                interpreter.evalNode(varAssignNode.getAssignmentExpr()));
        return "<var assign>";
    }
}

File: src/main/java/org/nyxlang/interpreter/visitor/VarAssignVisitor.kt
package org.nyxlang.interpreter.visitor

import org.nyxlang.interpreter.IInterpreter
import org.nyxlang.interpreter.result.IRuntimeResult
import org.nyxlang.interpreter.result.emptyResult
import org.nyxlang.parser.ast.INode
import org.nyxlang.parser.ast.VarAssignNode

/**
 * Interprets the var assign node.
 */
class VarAssignVisitor : IVisitor {
    override fun matches(node: INode) = VarAssignNode::class == node::class
    override fun visit(interpreter: IInterpreter, node: INode): IRuntimeResult {
        val varAssignNode = node as VarAssignNode
        val varName = varAssignNode.identifier
        val varAssignment = varAssignNode.assignmentExpr
        interpreter.activationRecord!![varName] = interpreter.evalNode(varAssignment).data
        return emptyResult()
    }
}